### PR TITLE
Adds snappy support for http_listener_v2

### PR DESCRIPTION
**Pull request by helenosheaa**
Adds snappy support plus test

Relates to #8682, #8967 - one use is for with prometheus remote write parsing

### CODE REVIEW

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 1]
```diff
@@ -268,6 +269,16 @@ func (h *HTTPListenerV2) collectBody(res http.ResponseWriter, req *http.Request)
...
+	// Handle snappy request bodies
+	if req.Header.Get("Content-Encoding") == "snappy" {
+		bytes, err = snappy.Decode(nil, bytes)
+		if err != nil {
+			h.Log.Debug(err.Error())
+			badRequest(res)
+			return nil, false
+		}
+	}
+
```
Comment (srebhan) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:281]: Is it possible to use the `snappy.NewReader` function and match the processing of `gzip`!?

**Review by srebhan — COMMENTED** [review]
I would prefer to match the pattern of gzip and construct a reader. Is there any reason to _not_ do it this way?

**Comment by helenosheaa** [comment]
Thanks for the review, I'll take a look at matching the gzip pattern

**Comment by helenosheaa** [comment]
So after just having another look at the docs, we want to use Decode rather than a newReader as we need the block format rather than streaming format (specifically for the prometheus remote write use case). Bit more context [here](https://pkg.go.dev/github.com/golang/snappy#:~:text=The%20block%20format%20is%20the,at%20the%20time%20compression%20starts.)

File: plugins/inputs/http_listener_v2/http_listener_v2.go (hunk 1 above)
Comment (helenosheaa) [reply plugins/inputs/http_listener_v2/http_listener_v2.go:281]: See comment below

Comment (srebhan) [reply plugins/inputs/http_listener_v2/http_listener_v2.go:281]: Could you maybe add a small comment here that we need the block-decode!?

Comment (helenosheaa) [reply plugins/inputs/http_listener_v2/http_listener_v2.go]: Yes, added

Comment (srebhan) [reply plugins/inputs/http_listener_v2/http_listener_v2.go]: Thanks!

**Review by srebhan — APPROVED** [review]
Looks good to me.

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 2]
```diff
@@ -268,6 +269,17 @@ func (h *HTTPListenerV2) collectBody(res http.ResponseWriter, req *http.Request)
 		return nil, false
 	}
 
+	// Handle snappy request bodies
+	if req.Header.Get("Content-Encoding") == "snappy" {
```
Comment (reimda) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:273]: About 20 lines above there's a similar block for gzip.  It looks like we should change the "if content-encoding == gzip" to a switch statement that also handles snappy, then move this new block into the switch statement.  The snappy decoding needs to happen at the same place gzip does, namely before the MaxBytesReader.

**Review by reimda — CHANGES_REQUESTED** [review]

File: plugins/inputs/http_listener_v2/http_listener_v2.go (hunk 2 above)
Comment (helenosheaa) [reply plugins/inputs/http_listener_v2/http_listener_v2.go:273]: The NewReader function of snappy would return a reader such as the gzip implementation so the switch would work but the block implementation of snappy that we need (for prometheus remote write) is a Decode which takes in bytes and returns bytes. So I think it needs to be after the reader on 266 unless I'm missing something.

File: plugins/inputs/http_listener_v2/http_listener_v2.go [hunk 3]
```diff
@@ -268,6 +269,17 @@ func (h *HTTPListenerV2) collectBody(res http.ResponseWriter, req *http.Request)
...
+		// Requires snappy decode for block format
+		bytes, err = snappy.Decode(nil, bytes)
```
Comment (ssoroka) [inline plugins/inputs/http_listener_v2/http_listener_v2.go:275]: Let's use snappy.NewReader() instead which works on streams and not byte slices, then  ioutil.ReadAll is the last step of both cases, so definitely go with the switch statement

File: plugins/inputs/http_listener_v2/http_listener_v2_test.go [hunk 4]
```diff
@@ -327,6 +328,37 @@ func TestWriteHTTPGzippedData(t *testing.T) {
...
+	require.NoError(t, err)
+	req.Header.Set("Content-Encoding", "snappy")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	require.NoError(t, err)
+	require.EqualValues(t, 204, resp.StatusCode)
+
+	hostTags := []string{"server01"}
+	acc.Wait(len(hostTags))
```
Comment (ssoroka) [inline plugins/inputs/http_listener_v2/http_listener_v2_test.go:352]: should this just be 1, since it's 1 metric?

**Review by ssoroka — CHANGES_REQUESTED** [review]

File: plugins/inputs/http_listener_v2/http_listener_v2.go (hunk 3 above)
Comment (helenosheaa) [reply plugins/inputs/http_listener_v2/http_listener_v2.go:278]: As discussed refactored to use a switch statement, but sticking with Decode as the NewReader doesn't support the block format

**Review by ssoroka — APPROVED** [review]

**Pull request merged**